pkg/dispers/conceptindexor.go

```diff
@@ -0,0 +1,200 @@
+package enclave
+
+import (
+	"errors"
+
+	"golang.org/x/crypto/scrypt"
+
+	"github.com/cozy/cozy-stack/pkg/couchdb"
+	"github.com/cozy/cozy-stack/pkg/couchdb/mango"
+	"github.com/cozy/cozy-stack/pkg/crypto"
+	"github.com/cozy/cozy-stack/pkg/dispers/dispers"
+	"github.com/cozy/cozy-stack/pkg/prefixer"
+)
+
+const (
+	defaultN       = 32768
+	defaultR       = 8
+	defaultP       = 1
+	defaultDkLen   = 32 // hash length
+	defaultLenSalt = 128
+
+	doctypeSalt = "io.cozy.hashconcept"
+)
+
+var prefixerCI = prefixer.ConceptIndexorPrefixer
+
+// ConceptDoc is used to save a concept's salt into Concept Indexor's database
+type ConceptDoc struct {
+	ConceptID  string `json:"_id,omitempty"`
+	ConceptRev string `json:"_rev,omitempty"`
+	Concept    string `json:"concept,omitempty"`
+	Salt       string `json:"salt,omitempty"`
+}
+
+// ID is used to get ID
+func (t *ConceptDoc) ID() string {
+	return t.ConceptID
+}
+
+// Rev is used to get Rev
+func (t *ConceptDoc) Rev() string {
+	return t.ConceptRev
+}
+
+// DocType is used to get DocType
+func (t *ConceptDoc) DocType() string {
+	return doctypeSalt
+}
+
+// Clone is used to create another ConceptDoc from this ConceptDoc
+func (t *ConceptDoc) Clone() couchdb.Doc {
+	cloned := *t
+	return &cloned
+}
+
+// SetID is used to set the Doc ID
+func (t *ConceptDoc) SetID(id string) {
+	t.ConceptID = id
+}
+
+// SetRev is used to set Rev
+func (t *ConceptDoc) SetRev(rev string) {
+	t.ConceptRev = rev
+}
+
+// addSalt does not check if one salt is already existing
+// It only creates a doc in ConceptIndexor's database
+func addSalt(concept string) error {
+
+	ConceptDoc := &ConceptDoc{
+		ConceptID:  "",
+		ConceptRev: "",
+		Concept:    concept,
+		Salt:       string(crypto.GenerateRandomBytes(defaultLenSalt)),
+	}
+	return couchdb.CreateDoc(prefixerCI, ConceptDoc)
+}
+
+func getSalt(concept string) (string, error) {
+
+	// Create index to run a mango query
+	salt := ""
+	err := couchdb.DefineIndex(prefixerCI, mango.IndexOnFields(doctypeSalt, "concept-index", []string{"concept"}))
```

Comment:
What happens if the index is already declared? This should probably not be here but in a init section, like the stack does when started.

Reply:
Ok. I find the place and I move it there.